clients/js/src/compat.ts

```diff
@@ -0,0 +1,289 @@
+import { KeyManagerSignedPublicKey } from '@oasisprotocol/client-rt/dist/types.js';
+
+import { EthCall, prepareSignedCall } from './signed_calls.js';
+import {
+  MAINNET_CHAIN_ID,
+  MAINNET_RT_ID,
+  TESTNET_CHAIN_ID,
+  TESTNET_RT_ID,
+} from './index.js';
+
+const TESTNET_RPC = 'https://grpc.oasis.dev';
+const MAINNET_RPC = 'https://testnet.grpc.oasis.dev';
+
+const PK_CACHE: Map<number /* chain id */, KeyManagerSignedPublicKey> =
+  new Map();
+
+export type UpstreamProvider =
+  | { request: (args: Web3ReqArgs) => Promise<unknown> } // EIP-1193
+  | { sendAsync: AsyncSend } // old MetaMask
+  | { send: AsyncSend } // Web3.js
+  | (EthersSigner & TypedDataSigner);
+
+type AsyncSend = (
+  args: Web3ReqArgs,
+  cb: (err: unknown, ok?: unknown) => void,
+) => void;
+
+interface Web3ReqArgs {
+  readonly method: string;
+  readonly params?: readonly unknown[] | object;
+}
+
+function hookEthersCall<T>(
+  signer: EthersSigner & TypedDataSigner,
+  call: EthersCall<T>,
+): EthersCall<T> {
+  return async (callP, ...rest) => {
+    return call(
+      (await callNeedsSigning(callP))
+        ? await prepareSignedCall(undefer(callP) as unknown as EthCall, signer)
+        : callP,
+      ...rest,
+    );
+  };
+}
+
+async function callNeedsSigning(
+  callP: Deferrable<TransactionRequest>,
+): Promise<boolean> {
+  const [from, to] = await Promise.all([callP.from, callP.to]);
+  return to !== undefined && from !== undefined && !/^(0x)?0{40}$/.test(from);
+}
+
+type EthersCall<T> = (
+  tx: Deferrable<TransactionRequest>,
+  ...rest: unknown[]
+) => Promise<T>;
+
+type Deferrable<T> = {
+  [K in keyof T]: T[K] | Promise<T[K]>;
+};
+
+async function undefer<T>(obj: Deferrable<T>): Promise<T> {
+  return Object.fromEntries(
+    await Promise.all(Object.entries(obj).map(async ([k, v]) => [k, await v])),
+  );
+}
+
+type ProviderOptions = Partial<{
+  /** The custom Sapphire GRPC endpoint. */
+  rpc: string;
+  runtimeId: string;
+}>;
+
+export function wrapSigner<T extends UpstreamProvider>(
+  upstream: T,
+  opts?: ProviderOptions,
+): T {
+  if (EthersSigner.isSigner(upstream) && '_signTypedData' in upstream) {
+    const signer: EthersSigner & TypedDataSigner = upstream;
+    const provider = signer.provider;
+    if (provider === undefined) {
+      throw new Error(
+        'ethers Signer must be connected to a Provider before wrapping',
+      );
+    }
+    return signer.connect(
+      new Proxy(provider as EthersProvider, {
+        get(target, prop) {
+          if (prop === 'sendTransaction') {
+            return (async (signedTx) => {
+              const [ptTx, chainId] = await Promise.all([
+                signedTx,
+                signer.getChainId(),
+              ]);
+              const [encTx] = await encrypt(ptTx, chainId, opts);
+              return provider.sendTransaction(encTx);
+            }) as EthersProvider['sendTransaction'];
```

Comment:
The wrapper logic doesn't sound correct to me. It seems that we should encrypt the transaction data first and then sign it and finally send it out. Shouldn't we Proxy `EthersSigner['sendTransation']` instead of `EthersProvider['sendTransaction']` here?

Reply:
> Shouldn't we proxy [...]

Encryption happens at the lowest layer (the provider) because we intend to make it possible for proxies in the middle of the exchange to encrypt the transaction.

Reply:
This isn't true for signed queries, though, which I'd agree is odd, but it's weird to have something like:

```js
const origCall  = {
    gas: 42,
    gasLimit: 1000,
    from: '0xabc...',
    to: '0xdef..',
    data: '0xffff',
};
const encCall = {
  data: enc(origCall.data),
  ...origCall,
};
const signedCall = {
  data: encode({
    data: encCall.data,
    signature: sign(encode(encCall))
  }),
  ...encCall
};
```

versus now where it's
```
const signedCall = {
  call: encCall,
  signature: sign(encode(encCall))
}
```
which is more normal in terms of signatures, but doesn't permit proxy encryption

Reply:
> Encryption happens at the lowest layer (the provider) because we intend to make it possible for proxies in the middle of the exchange to encrypt the transaction.

I just didn't understand how can we encrypt a signed transaction, which would invalidate the signatures. 

https://github.com/ethers-io/ethers.js/blob/01aea705ce60b1c42d2f465b162cb339a0e94392/packages/abstract-signer/src.ts/index.ts#L125

Reply:
> 
```
const signedCall = {
  data: encode({
    data: encCall.data,
    signature: sign(encode(encCall))
  }),
  ...encCall
};
```

This format sounds more natural to me. The second format is also fine since we have already handled that in our evm module. The format here doesn't really matter since the `data` field in a call request doesn't involve any sign process at the wallet level.

Reply:
https://testnet.explorer.sapphire.oasis.dev/tx/0x7ab6527a057c486fc327ef6ca65a34b5d1b40f360452b88146736aaad4a6923c/token-transfers

I changed the proxy to Signer['sendTransaction'] and use 'CALLFORMAT_PLAIN', it deployed successfully.

```
  if (EthersSigner.isSigner(upstream) && '_signTypedData' in upstream) {
    const signer: EthersSigner & TypedDataSigner = upstream;
    const provider = signer.provider;
    if (provider === undefined) {
      throw new Error(
        'ethers Signer must be connected to a Provider before wrapping',
      );
    }
    return new Proxy(signer as EthersSigner, {
      get(target, prop) {
	 if (prop === 'sendTransaction') {
	   return (async (originTx) => {
	     const chainId = await signer.getChainId();
             const [encTx] = await encrypt(originTx.data as string, chainId, opts);
	     originTx.data = encTx;
             return signer.sendTransaction(originTx);	     
	   }) as EthersSigner['sendTransaction'];
	 }
	let value = Reflect.get(target, prop);
	return typeof value == 'function' ? value.bind(target) : value;
      }
    }) as T;
  }
```

Reply:
> I just didn't understand how can we encrypt a signed transaction, which would invalidate the signatures.

Encryption doesn't invalidate signatures. You just need to unwrap the encryption before verifying, if needed.